design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,210 @@
+- The agent maintains a mapping of the types and their corresponding plugins.
+- When an an operation is received for a type, it is delegated to the corresponding plugin.
+
+### Log Management
+
+- `tedge-agent` to add support for multiple log management plugins defined under `/etc/tedge/log-plugins`.
+- Migrate the existing file-based log functionality into a dedicated `file` plugin which is available by default.
+- To retrieve logs from other sources like systemd journal or docker, add respective plugins for these sources.
+- Each plugin must support the following sub-commands:
+  - `list`: list all the log types that it supports.
+    Used to detect if this is a valid plugin if it exits with exit code 0.
+    The output is used only when auto-discovery for log types is turned ON for this plugin.
+  - `get <type> <temp-log-file-path> [--from <timestamp>] [--to <timestamp>] [--filter <filter-text>]`:
+    Used to fetch the log for the given type within the provided log range.
+    The log content must be written to the temporary file passed provided by the agent.
+- The plugin directory would also have a `conf.d` directory where each plugin can store their configuration in a toml file.
+  These config files are used to turn type discovery on/off and to define inclusion/exclusion list when type discovery is turned on:
+  ```docker.toml
+  [docker]
+  auto_discover=true
+
+  [[docker.include]]
+  pattern = "nginx"
+
+  [[docker.exclude]]
+  pattern = "kube-*"
+  ```
+- The `auto_discover` and `include`/`exclude` configs are used by the `tedge-agent`
+  to process the supported types listed by that plugin.
+  Any plugin-specific settings can also be defined in these files, which are ignored by the agent,
+  but can be used by the plugin.
+- The main config file for a plugin must be named after the plugin itself (e,g: `docker.sh` -> `docker.toml`).
+  Key configs like `auto_discover` can only be defined in this main file.
+  The inclusion/exclusion list can be defined in extension files as well,
+  but the table names with the plugin name prefix (e.g: `[docker.include]` and `[docker.exclude]`) must be used in those as well.
+- When new software is installed, their corresponding `include/exclude` entries can be appended to the main plugin config itself,
+  or created in an independent extension file.
+
+The agent uses the plugins as follows:
+
+- On startup, gather the supported types from all the plugins by running the `list` command on them.
+  Those failing to execute the `list` command are not qualified as valid plugins and ignored.
+- Whenever the `/etc/tedge/log-plugins` directory is updated (new plugin installed or existing one removed),
+  agent refreshes the supported types.
+  Simply touching this directory would also trigger a refresh.
+- When the supported types are published over mqtt, their source plugin information is also appended to that type
+  in the format <log_type>::<plugin_type>.
```

Comment:
I'm wondering if we can address the problem that users face if they have a heterogeneous fleet, and only care about getting the "mosquitto" logs regardless if it is using the file-based plugin or the systemd. By requiring the "::<type>" suffix when using any other plugin apart from the "file" based plugin, it would result in users having to use different command depending on the device.

Is there anyway where we can try to resolve the log type without the provider...and if there is only one match, then it is used, otherwise, the first match is used...if the user wants to be more specific, then can provide the `::<plugin>` info...though this would complicate the publishing of the supported types.

Reply:
This is a tricky one. It sounds doable as you've proposed with some additional checks while log types from different plugins are merged. But I'm just wondering if this inconsistency where the types coming from the same plugin appear differently in the cloud, some with the suffix and the others without it, would be more confusing to the user as they wouldn't know which plugin is responsible for that type.